book/src/developers/designs/identity-verification.md

```diff
@@ -0,0 +1,160 @@
+## Draft proposal for and Identity verification system with an out of band challenge
+
+Implements #337
+
+### Goal:
+
+Providing a way to allow 2 subject to authenticate each other without prior knowledge of each other
```

Comment:
Well, yes and no. The idea is to validate that a person you are communicating with is exactly who they say they are. 

For example, if someone wants to you transfer a bunch of money, then you can phone them and validate who they are, with kanidm acting as the assessor that both parties can firstly login as who they claim to be, and that they have active and current access to that system indicating a valid session.

It's effectively a way for a person to remotely validate that a person can authenticate to Kanidm, but without having to actually invasively check that they are able to. 

So with this in mind we need:

* Something that allows both sides to validate the other. We don't want person A to read our their code and B "lies" and just says yep that was valid. We need A to validate B and vice versa.
* This means that codes have to be easy to read and type
* The codes have to work for enough time for someone on a phone call to muck about and login and such.
* The codes have to also *change* after a period of time. 
* The codes must be unique to all pairs of people as they validate - A and B should have different codes to A and C or B and C.
* The codes must be consistent between all Kanidm servers due to replication segmentation and netsplits etc. 

This is why I suggested ecdh private key on each account that can then be used for every pair of accounts to arrive at a unique secret. 

We can then use that unique secret as a TOTP seed, since TOTP changes on a time schedule and is easy to read and type. The TOTP in Kanidm also has a "grace" allowing a single previous TOTP to be valid meaning that small de-sync in time steps is not noticed by the parties involved. 

The "trick" is that we can't have each side get the same totp. We want them to have a totp for A to B, and a seperate code for B to A. 

This means we need a deterministic way to take our derive secret and have each server do the same derivation for A and B to get their seperate it/out secrets.

AN example could be the account with the lower uuid takes uses the derived secret "as is" and the higher uuid reverses the byte array. This is a way both servers without communication can derive the same totp seeds without communication. 

Does that help?

Reply:
> 
> Does that help?

Yess now it  is indeed more clear! I didn't think about users possibly lying about the correctness of their TOTP, now I see how we need two TOTPs to allow users to mutually authenticate each other. For the time being, I'm thinking about something of the kind: 
First of all let's supposed Alice has the lowest UUID
- Alice's server computes the hash of the current time (rounded to 30 seconds) + her UUID, and encrypts it with her private key and Bob's pubic key
- Bob's server computes the hash of the current time (same 30s rounding) + his UUID reversed, and encrypts it with his private key and Alice's public key
- They exchange the encrypted messages by ie phone call (they will be in a human friendly format obv), they type them in their own servers and eventually the servers will verify the correctness of the TOTPs by decrypting them with the corresponding private and public key and checking whether the plaintext matches the secret computed locally. Note that the servers will have to take into account that the other server has inverted the UUID.  

I used the UUID strategy you suggested just to check if I understood, but I intend to think about it a little more to see if I can come up with something on my own! 
For example I was thinking we could compute the hash using both of their UUID while also maintaining the inversion for the highest UUID, ie: 
Alice computes hash(time, AliceUUID, BobUUID)
Bob computes hash(time, ReversedAliceUUID, ReversedBobUUID)
Both of them compute the hash using both of the UUIDs, but the account with the lowest UUID reverses the byte array of the UUIDs.  
I was also thinking we could derive encryption keys from one UUID and use them to hash the other UUID. 
Alice would derive an encryption key from her UUID and would use it to compute an HMAC of Bob's UUID, like HMAC(KDF(AliceUUID), BobUUID), Bob would be able to verify the correctness of the HMAC by computing it again locally.